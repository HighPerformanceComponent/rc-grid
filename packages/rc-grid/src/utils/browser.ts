// from https://github.com/ag-grid/ag-grid/blob/latest/community-modules/core/src/ts/utils/browser.ts

let browserScrollbarWidth: number
let invisibleScrollbar: boolean

function initScrollbarWidthAndVisibility(): void {
    const { body } = document
    const div = document.createElement('div')
    div.style.width = '100px'
    div.style.height = '100px'
    div.style.opacity = '0'
    div.style.overflow = 'scroll'
    ;(div.style as any).msOverflowStyle = 'scrollbar' // needed for WinJS apps
    div.style.position = 'absolute'

    body.appendChild(div)

    let width: number = div.offsetWidth - div.clientWidth

    // if width is 0 and client width is 0, means the DOM isn't ready
    if (width === 0 && div.clientWidth === 0) {
        width = undefined
    }

    // remove div
    if (div.parentNode) {
        div.parentNode.removeChild(div)
    }

    browserScrollbarWidth = width

    invisibleScrollbar = width === 0 || width === undefined
}

export function getScrollbarWidth(): number {
    if (browserScrollbarWidth === undefined) {
        initScrollbarWidthAndVisibility()
    }
    return browserScrollbarWidth
}

export function isInvisibleScrollbar(): boolean {
    if (invisibleScrollbar === undefined) {
        initScrollbarWidthAndVisibility()
    }
    return invisibleScrollbar
}

let clipboard: HTMLInputElement = null

/**
 * 写入剪贴板内容
 */
export const writeClipboardText = (text: string) => {
    if (clipboard === null) {
        clipboard = document.createElement("input");
        clipboard.style.cssText = 'position: absolute; top: -9999px;';
        document.body.appendChild(clipboard);
    }
    clipboard.value = text;
    clipboard.select();
    document.execCommand('copy');
}