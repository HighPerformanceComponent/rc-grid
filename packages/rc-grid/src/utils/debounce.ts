interface Option {
    // 最大等待时间
    maxTime?: number
}

const debounce = (fn: Function, time: number, option?: Option) => {
    // 最后一次调用的参数
    let lastArgs: any;

    let lastInvokeTime: number;
    let invokeSelf: any;
    
    // 调用的返回信息
    let invokeResult: any;

    let lastCallTime: number = 0
    
    let timeout: any;

    const invokeFunc = () => {
        lastCallTime = Date.now()
        const result = fn.apply(invokeSelf, lastArgs)
        invokeResult = result
    }

    const shouldInvoke = (currentTime: number) => {
        const lastInvokeDiff = currentTime - lastInvokeTime
        lastInvokeTime = currentTime

        if (timeout) {
            clearTimeout(timeout)
        }

        if (lastInvokeDiff >= time) {
            return true
        }
        
        if (option?.maxTime && currentTime - lastCallTime > option.maxTime ) {
            return true
        }

        timeout = setTimeout(() => {
            invokeFunc()
        }, time - lastInvokeDiff)

        return false
    }

    function debounced(...args: any[]) {
        invokeSelf = this
        lastArgs = args
        const isInvoking = shouldInvoke(Date.now())
        if (isInvoking) {
            invokeFunc()
        }
        return invokeResult 
    }

    return debounced
}

export default debounce