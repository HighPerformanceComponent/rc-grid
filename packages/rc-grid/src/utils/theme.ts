
export interface Theme {
    'grid-border': string
    'grid-text-color': string
    'grid-row-background-color': string
    'grid-row-background-color-select': string
    'grid-row-background-color:hover': string
    'grid-row-cell-border-right': string
    'grid-row-cell-border-bottom': string
    'grid-row-cell-background-color-readonly': string
    'grid-row-cell-select': string
    'grid-header-cell-border-right': string
    'grid-header-cell-border-bottom': string
    'grid-header-cell-background-color': string
}

export const dark: Theme = {
    'grid-border': '1px solid hsl(0deg 0% 40%)',
    'grid-text-color': '#fff',
    'grid-row-background-color': '#000',
    'grid-row-background-color-select': 'hsl(0deg 0% 30%)',
    'grid-row-background-color:hover': 'hsl(0deg 0% 30%)',
    'grid-row-cell-border-right': '1px solid hsl(0deg 0% 40%)',
    'grid-row-cell-border-bottom': '1px solid hsl(0deg 0% 40%)',
    'grid-row-cell-background-color-readonly': 'hsl(0deg 0% 20%)',
    'grid-row-cell-select': 'inset 0 0 0 1px #66afe9',
    'grid-header-cell-border-right': '1px solid hsl(0deg 0% 40%)',
    'grid-header-cell-border-bottom': '1px solid hsl(0deg 0% 40%)',
    'grid-header-cell-background-color': 'hsl(0deg 0% 15%)'
}