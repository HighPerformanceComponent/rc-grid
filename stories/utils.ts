import produce from 'immer'
import { Column } from '../src'

// eslint-disable-next-line import/prefer-default-export
export function onHeaderDrop<T>(
    columns: Column<T>[],
    source: Column<T>,
    target: Column<T>
) {
    const sourceIndex = columns.findIndex((ele) => ele.name === source.name)
    const targetIndex = columns.findIndex((ele) => ele.name === target.name)
    return produce(columns, (changeColumns) => {
        changeColumns.splice(sourceIndex, 1)
        changeColumns.splice(targetIndex, 0, source)
    })
}
