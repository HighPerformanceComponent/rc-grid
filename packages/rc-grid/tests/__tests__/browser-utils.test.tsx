import { isInvisibleScrollbar, writeClipboardText } from '../../src/utils/browser'

test('isInvisibleScrollbar test', () => {
    expect(isInvisibleScrollbar()).toMatchSnapshot()
})

test('writeClipboardText test', async () => {
    writeClipboardText('test 123');
    expect(document.body).toMatchSnapshot()
})