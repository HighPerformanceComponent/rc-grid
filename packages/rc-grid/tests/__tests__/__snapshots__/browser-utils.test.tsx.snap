// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isInvisibleScrollbar test 1`] = `true`;

exports[`writeClipboardText test 1`] = `
<body>
  <input
    style="position: absolute; top: -9999px;"
  />
</body>
`;
